
# Task 2. Problem-solving

## Project contains:
* [2.1. Download the event_data.csv dataset](#1-bullet)
* [2.2. Highlight user cohorts based on the week of registration in the application. ](#2-bullet)
* [2.3. How many unique users in the cohort with ID 33](#3-bullet)
* [2.4. Download the event_data.csv dataset](#4-bullet)
* [2.5. Highlight user cohorts based on the week of registration in the application. ](#5-bullet)
* [2.6. How many unique users in the cohort with ID 33](#6-bullet)
* [2.7. Download the event_data.csv dataset](#7-bullet)
* [2.8. Highlight user cohorts based on the week of registration in the application. ](#8-bullet)
* [2.9. How many unique users in the cohort with ID 33](#9-bullet)

<a id='1-bullet'></a>

### 2.1. Download the event_data.csv dataset

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('C:/Users/Iljam/Downloads/event_data.csv')

In [3]:
df.head()

user_id           event_date    event_type  purchase_amount
0  c40e6a  2019-07-29 00:02:15  registration              NaN
1  a2b682  2019-07-29 00:04:46  registration              NaN
2  9ac888  2019-07-29 00:13:22  registration              NaN
3  93ff22  2019-07-29 00:16:47  registration              NaN
4  65ef85  2019-07-29 00:19:23  registration              NaN

<a id='2-bullet'></a>

### 2.2. Highlight user cohorts based on the week of registration in the application.

In [4]:
df['event_date'] = pd.to_datetime(df['event_date'])

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 79742 entries, 0 to 79741
Data columns (total 4 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   user_id          79742 non-null  object        
 1   event_date       79742 non-null  datetime64[ns]
 2   event_type       79742 non-null  object        
 3   purchase_amount  6207 non-null   float64       
dtypes: datetime64[ns](1), float64(1), object(2)
memory usage: 2.4+ MB


In [6]:
df['cohort'] = df['event_date'].dt.week

<ipython-input-6-2b8da0700fea>:1: FutureWarning: Series.dt.weekofyear and Series.dt.week have been deprecated.  Please use Series.dt.isocalendar().week instead.
  df['cohort'] = df['event_date'].dt.week


In [7]:
df

user_id          event_date    event_type  purchase_amount  cohort
0      c40e6a 2019-07-29 00:02:15  registration              NaN      31
1      a2b682 2019-07-29 00:04:46  registration              NaN      31
2      9ac888 2019-07-29 00:13:22  registration              NaN      31
3      93ff22 2019-07-29 00:16:47  registration              NaN      31
4      65ef85 2019-07-29 00:19:23  registration              NaN      31
...       ...                 ...           ...              ...     ...
79737  930c23 2019-09-01 23:57:41  simple_event              NaN      35
79738  a84999 2019-09-01 23:57:50  simple_event              NaN      35
79739  175e4d 2019-09-01 23:59:40  simple_event              NaN      35
79740  1c2210 2019-09-01 23:59:51  simple_event              NaN      35
79741  d799dc 2019-09-01 23:59:54  simple_event              NaN      35

[79742 rows x 5 columns]

<a id='3-bullet'></a>

### 2.3 How many unique users in the cohort with ID 33

In [8]:
pd.Series(df[df['cohort'] == 33]['user_id'].unique()).count()

5102

Answer is 5102 unique users

<a id='4-bullet'></a>

### 2.4 For each event, highlight the indicator lifetime - the weekly lifetime of the cohort.

In [9]:
df['event_date_week'] = [int(i) + 1 for i in df['event_date'].dt.strftime('%W')]
df_reg = df[df['event_type'] == 'registration']
df_not_reg = df[df['event_type'] != 'registration']
df_reg = df_reg[['user_id', 'event_date_week']]
df_reg.columns = ['user_id', 'registration_week']

In [10]:
df_reg

user_id  registration_week
0      c40e6a                 31
1      a2b682                 31
2      9ac888                 31
3      93ff22                 31
4      65ef85                 31
...       ...                ...
79697  a7c20c                 35
79703  03e836                 35
79710  fb0e00                 35
79712  46b4f8                 35
79716  48b765                 35

[9996 rows x 2 columns]

In [11]:
df = df.merge(df_reg, how='left', on='user_id')
df['cohort_lifetime'] = df['event_date_week'] - df['registration_week']

In [12]:
df['cohort_lifetime'].unique()

array([0, 1, 2, 3, 4], dtype=int64)

In [13]:
df

user_id          event_date    event_type  purchase_amount  cohort  \
0      c40e6a 2019-07-29 00:02:15  registration              NaN      31   
1      a2b682 2019-07-29 00:04:46  registration              NaN      31   
2      9ac888 2019-07-29 00:13:22  registration              NaN      31   
3      93ff22 2019-07-29 00:16:47  registration              NaN      31   
4      65ef85 2019-07-29 00:19:23  registration              NaN      31   
...       ...                 ...           ...              ...     ...   
79737  930c23 2019-09-01 23:57:41  simple_event              NaN      35   
79738  a84999 2019-09-01 23:57:50  simple_event              NaN      35   
79739  175e4d 2019-09-01 23:59:40  simple_event              NaN      35   
79740  1c2210 2019-09-01 23:59:51  simple_event              NaN      35   
79741  d799dc 2019-09-01 23:59:54  simple_event              NaN      35   

       event_date_week  registration_week  cohort_lifetime  
0                   31                 31                0  
1                   31                 31                0  
2                   31                 31                0  
3                   31                 31                0  
4                   31                 31                0  
...                ...                ...              ...  
79737               35                 32                3  
79738               35                 33                2  
79739               35                 32                3  
79740               35                 33                2  
79741               35                 35                0  

[79742 rows x 8 columns]

<a id='5-bullet'></a>

### 2.5  Build a summary table of changes in the Retention Rate for cohorts depending on lifetime.

In [14]:
cohorts = df.pivot_table(index = ['registration_week','cohort_lifetime'], 
                         aggfunc = {'user_id':'nunique'}).reset_index() 
initial_users_count = cohorts[cohorts['cohort_lifetime'] == 0]
initial_users_count = initial_users_count[['registration_week','user_id']]
initial_users_count = initial_users_count.rename(columns={'user_id':'cohort_users'}) 

cohorts = cohorts.merge(initial_users_count,on='registration_week')
cohorts['retention'] = cohorts['user_id']/cohorts['cohort_users'] 
retention_pivot = cohorts.pivot_table(index='registration_week',
                                      columns='cohort_lifetime',
                                      values='retention',
                                      aggfunc='sum') 

In [15]:
retention_pivot

cohort_lifetime      0         1         2         3        4
registration_week                                            
31                 1.0  0.927595  0.629367  0.356962  0.15038
32                 1.0  0.929303  0.648053  0.361168      NaN
33                 1.0  0.924205  0.661125       NaN      NaN
34                 1.0  0.929078       NaN       NaN      NaN
35                 1.0       NaN       NaN       NaN      NaN

<a id='6-bullet'></a>

### 2.6 What is the 3 week retention rate for a cohort with ID 32? Give the answer in percent, rounded to 2 decimal places, inclusive.

In [16]:
round(retention_pivot[[2]].head(2).tail(1)*100,2)

cohort_lifetime        2
registration_week       
32                 64.81

<a id='7-bullet'></a>

### 2.7 Build a summary table of changes in the indicator ARPPU (Average Revenue Per Paying User) for cohorts depending on lifetime.

In [17]:
df['purchase_amount'] = df['purchase_amount'].fillna(0).astype('int64')
df['purchase_amount'].unique()

array([ 0, 10, 50, 30, 40, 20], dtype=int64)

In [18]:
arppu = pd.pivot_table(df, index = 'cohort', columns = "cohort_lifetime", values = "purchase_amount")
arppu

cohort_lifetime         0         1         2         3         4
cohort                                                           
31               1.543939       NaN       NaN       NaN       NaN
32               1.901507  2.761124       NaN       NaN       NaN
33               1.625976  2.824745  2.614001       NaN       NaN
34               1.651042  2.781262  2.667231  2.535211       NaN
35               1.896666  2.615876  2.676797  2.558235  2.398568

<a id='8-bullet'></a>

### 2.8 What is the 3-week ARPPU of a cohort with ID 31? Give the answer with a floating point number, rounded to 2 decimal places, inclusive.

In [19]:
round(arppu[[3]].head(1)*100,2)

cohort_lifetime   3
cohort             
31              NaN

<a id='9-bullet'></a>

### 2.9 What is the median time between user registration and first purchase? Give the answer in seconds (!) As an integer.

Maybe I haven't interpreted the question, but isn't this question means median of lifetime ? If it's was a real case during work, I think I'll ask a mentor / teamlead or another person in charge what is it. So , my way of solving this problem here:

In [20]:
int(df['cohort_lifetime'].median())

1